### PR TITLE
Update NanoUI

### PR DESCRIPTION
**Pull request by neersighted**
This PR ports the new NanoUI code from Baystation12. While mostly under the hood, this will enable me to update and refactor our existing NanoUIs, and later NanoUI 'everything' on /tg/.

So far I have ported the Baycode, updated the libraries, refactored the JSON code, all the interfaces, implemented asset_cache, and re-factored the NanoUI API.

Eventually I will write a guide and NanoUI all the things -- however, these will be in separate PRs.

Some recent screenshots (unsorted):

### CODE REVIEW

**Comment by YotaXP** [comment]
You beautiful person, you.  Can we get a screenshot of any new UIs?

**Comment by neersighted** [comment]
This is mostly under the hood, and I'm off to dinner. However, I have this...
Air Alarm

**Comment by Anonus** [comment]
I notice some images for syndicate interfaces, would that mean upgrading the syndicate telecrystal consoles and shuttle movement computer might also happen?

**Comment by optimumtact** [comment]
:+1: on the condition that you make the json reader/writer procs fully typed again, as per our coding stnadards.

**Comment by Shadowlight213** [comment]
:+1:×10000000 so much yes.

File: code/modules/nano/JSON Reader.dm [hunk 1]
```diff
@@ -20,185 +20,186 @@
...
-		val		= ""
-	while(++i <= lentext(json))
-		var/char = get_char()
-		if(escape)
-			switch(char)
-				if("\\", "'", "\"", "/", "u")
-					val += char
+	proc
+		// scanner
+		ScanJson(json)
```
Comment (duncathan) [inline code/modules/nano/JSON Reader.dm]: Yeah definitely fix the relative pathing here, like oranges pointed out

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 2]
```diff
@@ -795,6 +795,7 @@ var/list/ai_list = list()
 
 	if(stat == 2)
 		return //won't work if dead
+
```
Comment (Shadowlight213) [inline code/modules/mob/living/silicon/ai/ai.dm]: Why is there a random break here reeee

File: code/ATMOSPHERICS/components/unary_devices/cryo.dm [hunk 3]
```diff
@@ -203,22 +203,18 @@
...
-
-	if(href_list["close"])
-		if(close_machine() == usr)
-			var/datum/nanoui/ui = SSnano.get_open_ui(usr, src, "main")
-			ui.close()
-			on = 1
 	if(href_list["switchOff"])
 		on = 0
 
+	if(href_list["ejectOccupant"])
```
Comment (optimumtact) [inline code/ATMOSPHERICS/components/unary_devices/cryo.dm]: One other thing, did you test cyro still works after these topic changes?

Comment (neersighted) [inline code/ATMOSPHERICS/components/unary_devices/cryo.dm]: Yep, checked it and it works great. Only tested with a mindless `/mob/living/carbon/human` though.

File: code/modules/mob/living/silicon/ai/ai.dm (hunk 2 above)
Comment (neersighted) [inline code/modules/mob/living/silicon/ai/ai.dm]: Because whitespace damnit.

Comment (optimumtact) [inline code/modules/mob/living/silicon/ai/ai.dm]: agreed

**Comment by Aranclanos** [comment]
>nanoui pr
>is not removing it

:-1:

**Comment by Iamgoofball** [comment]
fuck off

:+1:

**Comment by Supermichael777** [comment]
hope this fixes my weird clientside issues with nano in 508+

**Comment by Supermichael777** [comment]
uncheck file dr

**Comment by neersighted** [comment]
@Supermichael777 It's pretty close. Maybe by tomorrow?

**Comment by MrStonedOne** [comment]
Seems fine to me.

**Comment by MrStonedOne** [comment]
So far no bugs reported

This can be merged

**Comment by Shadowlight213** [comment]
It's happening!

**Pull request merged**